### PR TITLE
Refactor gRPC out of Webjobs.Script

### PR DESCRIPTION
**Pull request by ankitkumarr**
The change is to decouple `Webjobs.Script` with any of the gRPC stuff. This way only the `WebJobs.Script.Grpc` has the actual Grpc implementation and dependencies.

Fixes https://github.com/Azure/azure-functions-host/issues/5419

### CODE REVIEW

File: src/WebJobs.Script.Grpc/GrpcServiceCollectionsExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+namespace Microsoft.Azure.WebJobs.Script.Grpc
+{
+    public static class GrpcServiceCollectionsExtensions
+    {
+        public static IServiceCollection AddGrpc(this IServiceCollection services)
+        {
+            services.AddSingleton<FunctionRpc.FunctionRpcBase, FunctionRpcService>();
+            services.AddSingleton<IRpcServer, GrpcServer>();
+            services.AddSingleton<IRpcWorkerProcessFactory, RpcWorkerProcessFactory>();
+            services.AddSingleton<IRpcWorkerChannelFactory, RpcWorkerChannelFactory>();
```
Comment (ankitkumarr) [inline src/WebJobs.Script.Grpc/GrpcServiceCollectionsExtensions.cs]: I think I should refactor this such that in here, we only register `IRpcWorkerChannelFactory` (as discussed offline) and maybe also move `RpcInitializationService` here.

That way `WebJobs.Script.Grpc` only exports a channel factory and manages whatever service it needs internally.

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs [hunk 2]
```diff
@@ -216,12 +213,8 @@ public static void AddCommonServices(IServiceCollection services)
             // Add Language Worker Service
             // Need to maintain the order: Add RpcInitializationService before core script host services
             services.AddManagedHostedService<RpcInitializationService>();
```
Comment (ankitkumarr) [inline src/WebJobs.Script/ScriptHostBuilderExtensions.cs]: This should be moved to Grpc as well I think

**Comment by ankitkumarr** [comment]
@fabiocav, the tests are updated and the build looks good, if you want to have a look :)

File: src/WebJobs.Script/Environment/EnvironmentExtensions.cs [hunk 3]
```diff
@@ -9,7 +9,7 @@
 
 namespace Microsoft.Azure.WebJobs.Script
 {
-    internal static class EnvironmentExtensions
+    public static class EnvironmentExtensions
```
[the remark this answers is not in the document]
Comment (ankitkumarr) [reply src/WebJobs.Script/Environment/EnvironmentExtensions.cs]: Yes because `GrpcWorkerChannel` in `WebJobs.Script.Grpc` needs access to it.

**Comment by ankitkumarr** [comment]
@pragnagopa, friendly ping.

File: src/WebJobs.Script/ManagedDependencies/ManagedDependencyOptions.cs [hunk 4]
```diff
@@ -1,10 +1,6 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the MIT License. See License.txt in the project root for license information.
 
-using System.Collections.Generic;
-using Microsoft.Azure.WebJobs.Script.Grpc.Messages;
-using Newtonsoft.Json;
-
```
Comment (pragnagopa) [inline src/WebJobs.Script/ManagedDependencies/ManagedDependencyOptions.cs]: nit: undo as this is not related and required.

File: src/WebJobs.Script/Workers/ProcessManagement/IProcessRegistry.cs [hunk 5]
```diff
@@ -6,7 +6,7 @@
 
 namespace Microsoft.Azure.WebJobs.Script.Workers
 {
-    internal interface IProcessRegistry
+    public interface IProcessRegistry
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/ProcessManagement/IProcessRegistry.cs]: leave these as internal. If Grpc project needs them, add InternalsVisible to flag

File: src/WebJobs.Script/Workers/ProcessManagement/WorkerProcess.cs [hunk 6]
```diff
@@ -162,9 +162,9 @@ private void OnOutputDataReceived(object sender, DataReceivedEventArgs e)
             }
         }
 
-        internal abstract void HandleWorkerProcessExitError(WorkerProcessExitException langExc);
+        public abstract void HandleWorkerProcessExitError(WorkerProcessExitException langExc);
 
-        internal abstract void HandleWorkerProcessRestart();
+        public abstract void HandleWorkerProcessRestart();
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/ProcessManagement/WorkerProcess.cs]: Same here. Leave them as internal

File: test/WebJobs.Script.Tests/Workers/Rpc/TestFunctionRpcService.cs [hunk 7]
```diff
@@ -5,7 +5,7 @@
 using System.Collections.Generic;
 using System.Reactive.Linq;
 using Microsoft.Azure.WebJobs.Script.Eventing;
-using Microsoft.Azure.WebJobs.Script.Eventing.Rpc;
+using Microsoft.Azure.WebJobs.Script.Grpc.Eventing;
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Workers/Rpc/TestFunctionRpcService.cs]: Please move tests to a difference project as well in a separate PR.

File: .gitignore [hunk 8]
```diff
@@ -18,7 +18,7 @@ package-lock.json
 
 /packages
 tools/ExtensionsMetadataGenerator/packages
-/TestResults
```
[the remark this answers is not in the document]
Comment (ankitkumarr) [reply .gitignore]: I intentionally did this. `/TestResults` ignores files only from the root, whereas `TestResults/` takes care of sub-directories too. We have nested `TestResults` that pollute git status and don't need to be tracked.

File: src/WebJobs.Script/ManagedDependencies/ManagedDependencyOptions.cs (hunk 4 above)
Comment (ankitkumarr) [reply src/WebJobs.Script/ManagedDependencies/ManagedDependencyOptions.cs]: Removing `using Microsoft.Azure.WebJobs.Script.Grpc.Messages;` is required, for separating the project. I removed others too that weren't being used. I can add them back.

Comment (pragnagopa) [reply src/WebJobs.Script/ManagedDependencies/ManagedDependencyOptions.cs]: thanks..no need to add unsed statements back!

File: src/WebJobs.Script/Workers/ProcessManagement/IProcessRegistry.cs (hunk 5 above)
Comment (ankitkumarr) [reply src/WebJobs.Script/Workers/ProcessManagement/IProcessRegistry.cs]: Thanks, that's a good idea. I just updated the PR.

File: test/WebJobs.Script.Tests/Workers/Rpc/TestFunctionRpcService.cs (hunk 7 above)
Comment (ankitkumarr) [reply test/WebJobs.Script.Tests/Workers/Rpc/TestFunctionRpcService.cs]: Good call, I can make a different PR altogether after this new grpc work item. I will add an item to the Epic tracking this move.

Comment (ankitkumarr) [reply test/WebJobs.Script.Tests/Workers/Rpc/TestFunctionRpcService.cs]: On second thought, we do keep all our tests in the same project, but separate them per the "type" of the test. This may go against that pattern?

Also, this may be one of the only file(s) that could be moved. This project would still need to reference `Webjobs.Script.Grpc` in order for other tests to work properly.

Comment (pragnagopa) [reply test/WebJobs.Script.Tests/Workers/Rpc/TestFunctionRpcService.cs]: Go ahead a open a new issue to refactor the tests. If it does require lot of refactoring to get the tests working, no action needed.

File: src/WebJobs.Script/Workers/ProcessManagement/WorkerProcess.cs [hunk 9]
```diff
@@ -12,7 +12,7 @@
 
 namespace Microsoft.Azure.WebJobs.Script.Workers
 {
-    internal abstract class WorkerProcess : IWorkerProcess, IDisposable
+    public abstract class WorkerProcess : IWorkerProcess, IDisposable
```
Comment (pragnagopa) [inline src/WebJobs.Script/Workers/ProcessManagement/WorkerProcess.cs]: does this have to be public?

**Comment by fabiocav** [comment]
I know we've discussed a lot of the changes here. I'll be looking through this ASAP.

**Comment by fabiocav** [comment]
@ankitkumarr can we retarget this to `dev` and make any required changes? would be good to have this merged in as soon as possible.

**Comment by fabiocav** [comment]
Closing stale PRs. Please reopen if still relevant.
